enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,316 @@
+that communicates with the container-runtime
+at various lifecycle stages of container e.g. CreateContainer, UpdateContainer, etc.,
+and modifies the container's cgroup to include a set of shared CPUs.
+
+At first,
+shared CPUs will be part of Kubelet [reservedSystemCpus](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list).
+The reason to utilize Kubelet `reservedSystemCpus` is because of CPU manager, which is not aware of the CPUs
+lying in this pool, so it doesn't undo or changes the allocation logic performed by the plugin.
+Therefore, the plugin assigns the shared CPUs to containers without racing/conflict
+the CPU manager behavior. 
+More about this decision at the [Alternative](mixed-cpu-node-plugin.md#alternatives) section
+
+There is a plan for a feature called shared-partition that implements additional pool.
+After the shared-partition lands, the pool layout should look roughly like:
+1. `reservedSystemCpus`
+2. `garanteedCpus` <- new pool
+3. `sharedCpus` = all CPUs - `reservedSystemCpus` - `garanteedCpus`
+
+And mixed-cpu-node-plugin will allocate the shared cpus from the `sharedCpus` pool and not
+from the reservedSystemCpus anymore.
+
+The node-plugin populates a special device named `openshift.io/shared-cpus` to provide way for pods to request
+for this special type of CPUs.
+There's no meaning to the value/number of `openshift.io/shared-cpus` devices that the pod requests.
+For example:
+```yaml
+requests:
+   openshift.io/shared-cpus: 1
+```
+Or
+```yaml
+requests:
+   openshift.io/shared-cpus: 2
+```
+Are the same in that context, the container has access to the same number of shared-cpus in both cases.
+In other words, the device request only uses as a hint to tell the plugin that shared-cpus required for the container and
+does not indicate any actual value.
+
+The reason for specifying a device, and not, for example, an annotation,
+is because when application pod requests for a device, the scheduler makes the pod pending till device up and running.
+This gives the node plugin room for setup and not being depended on pod admission order.
+So only when the node plugin finishes the registration process with NRI and device plugin, the scheduler
+admits the application pod.
+
+Once a container under the pod requests and gets access to shared cpus, new environment variable named
+`OPENSHIFT_SHARED_CPUS=<CPU-IDs>` will be present under the container's environment variables.
+
+The mixed-cpu-node-plugin itself is essentially a pod that runs as a DaemonSet on the worker nodes.
+cluster-node-tuning-operator (NTO) will be modified to support the deployment of the DaemonSet and additional
+resources (ServiceAccount, RBAC resources, SecurityContextConstraint, etc.)
+NTO will also be responsible to watch, monitor and report the node-plugin's DaemonSet status.
+
+In addition, since both the components are related, mixed-cpu-node-plugin code would be vendor under NTO.
```

Comment:
@jmencak  @MarSik  @Tal-or  this means we will have to vendor stable branches/tags/(commits?) of this repo for every fix/bulk of fixes  once in a while to NTO as we discussed and agreed.